types/src/trusted_state.rs

```diff
@@ -73,23 +73,30 @@ impl TrustedState {
         }
     }
 
-    /// Create an initial trusted state from an epoch change ledger info.
-    pub fn from_epoch_change_ledger_info(epoch_change_li: &LedgerInfo) -> Result<Self> {
+    /// Create an initial trusted state from an epoch change ledger info and
+    /// a version inside that epoch.
+    pub fn from_epoch_change_ledger_info(
+        latest_version: Version,
+        epoch_change_li: &LedgerInfo,
+    ) -> Result<Self> {
+        ensure!(
+            latest_version != epoch_change_li.version(),
+            "A client can only enter an epoch on the boundary; only with a version inside that epoch",
+        );
+        ensure!(
+            latest_version > epoch_change_li.version(),
+            "The given version must be inside the epoch",
+        );
+
+        let validator_set = epoch_change_li.next_validator_set().ok_or_else(|| {
+            format_err!("No ValidatorSet in LedgerInfo; it must not be on an epoch boundary")
+        })?;
+
         // Generate the EpochInfo from the new validator set.
         let epoch_info = EpochInfo {
             epoch: epoch_change_li.epoch() + 1,
-            verifier: Arc::new(
-                epoch_change_li
-                    .next_validator_set()
-                    .ok_or_else(|| {
-                        format_err!(
-                            "No ValidatorSet in LedgerInfo; it must not be on an epoch boundary"
-                        )
-                    })?
-                    .into(),
-            ),
+            verifier: Arc::new(validator_set.into()),
         };
-        let latest_version = epoch_change_li.version();
```

Comment:
why not just change this to "+1".. it's for test only anyways, am I right?

Reply:
yeah, it's only for tests right now. i'm thinking how i can use `StartupInfo` to also bootstrap this, but i'll handle that later. I would prefer to keep it this way for now though, since you could start from a saved state somewhere inside an epoch and it would be undesirable to start only at the start of the epoch.